mllib/src/main/scala/org/apache/spark/mllib/stat/test/AndersonDarlingTest.scala

```diff
@@ -0,0 +1,289 @@
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.mllib.stat.test
+
+import scala.annotation.varargs
+
+import collection.immutable.ListMap
+
+import org.apache.commons.math3.distribution.{ExponentialDistribution, GumbelDistribution,
+  LogisticDistribution, NormalDistribution, WeibullDistribution}
+
+import org.apache.spark.Logging
+import org.apache.spark.rdd.RDD
+
+/**
+ * The Anderson-Darling (AD) test, similarly to the Kolmogorov-Smirnov (KS) test, tests whether the
+ * data follow a given theoretical distribution. It should be used with continuous data and
+ * assumes that no repeated values occur (the presence of ties can affect the validity of the test).
+ * The AD test provides an alternative to the KS test. Namely, it is better
+ * suited to identify departures from the theoretical distribution at the tails.
+ * It is worth noting that the the AD test's critical values depend on the
+ * distribution being tested against. The AD statistic is defined as
+ * {{{
+ * A^2 = -N - \frac{1}{N}\sum_{i = 0}^{N} (2i + 1)(\ln{\Phi{(x_i)}} + \ln{(1 - \Phi{(x_{N+1-i})})
+ * }}}
+ * where {{{\Phi}}} is the CDF of the given distribution and `N` is the sample size.
+ * For more information @see[[https://en.wikipedia.org/wiki/Anderson%E2%80%93Darling_test]]
+ */
+private[stat] object AndersonDarlingTest extends Logging {
+
+  object NullHypothesis extends Enumeration {
+    type NullHypothesis = Value
+    val OneSample = Value("Sample follows theoretical distribution.")
+  }
+
+  /**
+   * AndersonDarlingTheoreticalDist is a trait that every distribution used in an AD test must
+   * extend. The rationale for this is that the AD test has distribution-dependent critical values,
+   * and by requiring extension of this trait we guarantee that future additional distributions
+   * make sure to add the appropriate critical values (CVs) (or at least acknowledge
+   * that they should be added)
+   */
+  sealed trait AndersonDarlingTheoreticalDist extends Serializable {
+    val params: Seq[Double]  // parameters used to initialized the distribution
+
+    def cdf(x: Double): Double // calculate the cdf under the given distribution for value x
+
+    def getCVs(n: Double): Map[Double, Double] // return appropriate CVs, adjusted for sample size
```

Comment:
thanks for feedback. I've renamed and also renamed rawCVs to rawCriticalValues for consistency